plugins/connection/pct_remote.py

```diff
@@ -0,0 +1,403 @@
+        cli:
+          - name: private_key_file
+            option: "--private-key"
+      vmid:
+        description:
+          - Container ID
+        default: proxmox_vmid
+        vars:
+          - name: proxmox_vmid
+    notes:
+      - When NOT using this plugin as root, you need to have sudo installed on proxmox and setup so we can run it without prompting for the password.
+"""
+
+EXAMPLES = r"""
+# --------------------------------------------------------------
+# Setup sudo with passwordless access to pct for user 'ansible':
+# --------------------------------------------------------------
+# $ echo 'ansible ALL = (root) NOPASSWD: /usr/sbin/pct' > /etc/sudoers.d/ansible_pct
+#
+#
+# --------------------------------
+# Static inventory file: hosts.yml
+# --------------------------------
+# all:
+#   children:
+#     lxc:
+#       hosts:
+#         container-1:
+#           ansible_host: 10.0.0.10
+#           proxmox_vmid: 100
+#           ansible_connection: community.general.pct_remote
+#           remote_user: ansible
+#         container-2:
+#           ansible_host: 10.0.0.10
+#           proxmox_vmid: 200
+#           ansible_connection: community.general.pct_remote
+#           remote_user: ansible
+#     proxmox:
+#       hosts:
+#         proxmox-1:
+#           ansible_host: 10.0.0.10
+#
+#
+# ---------------------------------------------
+# Dynamic inventory file: inventory.proxmox.yml
+# ---------------------------------------------
+# plugin: community.general.proxmox
+# url: https://10.0.0.10:8006
+# validate_certs: false
+# user: ansible@pam
+# token_id: ansible
+# token_secret: !vault |
+#           $ANSIBLE_VAULT;1.1;AES256
+#           ...
+
+# want_facts: true
+# exclude_nodes: true
+# filters:
+#   - proxmox_vmtype == "lxc"
+# want_proxmox_nodes_ansible_host: false
+# compose:
+#   ansible_host: "'10.0.0.10'"
+#   ansible_connection: "'community.general.pct_remote'"
+#   remote_user: "'ansible'"
+#
+#
+# ----------------------
+# Playbook: playbook.yml
+# ----------------------
+---
+- hosts: lxc
+  tasks:
+    - debug:
+        msg: "This is coming from pct environment"
+"""
+
+from ansible import constants as C
+from ansible.errors import AnsibleError
+from ansible.plugins.connection.paramiko_ssh import Connection as SSH_Connection
+from ansible.utils.display import Display
+import uuid
+import os
+
+
+display = Display()
+
+
+def become_command():
+    """Helper function to get become_command """
+    return os.getenv('ANSIBLE_BECOME_METHOD', default=C.DEFAULT_BECOME_METHOD)
+
+
+class Connection(SSH_Connection):
+    ''' SSH based connections (paramiko) to Proxmox pct '''
+    transport = 'community.general.pct_remote'
+
+    def exec_command(self, cmd: str, in_data: bytes | None = None, sudoable: bool = True) -> tuple[int, bytes, bytes]:
+        ''' execute a command inside the proxmox container '''
+        cmd = ['/usr/sbin/pct', 'exec',
+               self.get_option('vmid'), '--', cmd]
```

Comment:
When adding user input into a command, you have to make sure it cannot do anything bad, for example by quoting it (https://docs.python.org/3/library/shlex.html#shlex.quote).

Reply:
done